src/backend/distributed/operations/shard_split.c

```diff
@@ -1145,3 +1221,650 @@ TryDropSplitShardsOnFailure(HTAB *mapOfShardToPlacementCreatedByWorkflow)
+			shardSplitSubscribersMetadataList);
+
+		/* 10) Wait for subscribers to catchup till source LSN */
+		XLogRecPtr sourcePosition = GetRemoteLogPosition(sourceConnection);
+		WaitForShardSplitRelationSubscriptionsToBeCaughtUp(sourcePosition,
+														   shardSplitSubscribersMetadataList);
+
+		/* 11) Create Auxilary structures */
+		CreateAuxiliaryStructuresForShardGroup(shardGroupSplitIntervalListList,
+											   workersForPlacementList);
+
+		/* 12) Wait for subscribers to catchup till source LSN */
+		sourcePosition = GetRemoteLogPosition(sourceConnection);
+		WaitForShardSplitRelationSubscriptionsToBeCaughtUp(sourcePosition,
+														   shardSplitSubscribersMetadataList);
+
+		/* 13) Block writes on source shards */
+		BlockWritesToShardList(sourceColocatedShardIntervalList);
+
+		/* 14) Wait for subscribers to catchup till source LSN */
+		sourcePosition = GetRemoteLogPosition(sourceConnection);
+		WaitForShardSplitRelationSubscriptionsToBeCaughtUp(sourcePosition,
+														   shardSplitSubscribersMetadataList);
+
+		/* 15) Drop Subscribers */
+		DropShardSplitSubsriptions(shardSplitSubscribersMetadataList);
+
+		/* 16) Drop Publications */
+		DropShardSplitPublications(sourceConnection, shardSplitHashMapForPublication);
+
+		/* 17) TODO(saawasek): Try dropping replication slots explicitly */
+
+		/*
+		 * 18) Drop old shards and delete related metadata. Have to do that before
+		 * creating the new shard metadata, because there's cross-checks
+		 * preventing inconsistent metadata (like overlapping shards).
+		 */
+		DropShardList(sourceColocatedShardIntervalList);
+
+		/* 19) Insert new shard and placement metdata */
+		InsertSplitChildrenShardMetadata(shardGroupSplitIntervalListList,
+										 workersForPlacementList);
+
+		/*
+		 * 20) Create foreign keys if exists after the metadata changes happening in
+		 * DropShardList() and InsertSplitChildrenShardMetadata() because the foreign
+		 * key creation depends on the new metadata.
+		 */
+		CreateForeignKeyConstraints(shardGroupSplitIntervalListList,
+									workersForPlacementList);
+
+		/* 21) Drop dummy shards.
+		 * TODO(saawasek):Refactor and pass hashmap.Currently map is global variable */
+		DropDummyShards();
+
+
+		/* 22) Close source connection */
+		CloseConnection(sourceConnection);
+
+		/* 23) Close all subscriber connections */
+		CloseShardSplitSubscriberConnections(shardSplitSubscribersMetadataList);
+
+		/* 24) Close connection of template replication slot */
+		CloseConnection(templateSlotConnection);
+	}
+	PG_CATCH();
+	{
+		/* Do a best effort cleanup of shards created on workers in the above block */
+		TryDropSplitShardsOnFailure(mapOfShardToPlacementCreatedByWorkflow);
+
+		/*TODO(saawasek): Add checks to open new connection if sourceConnection is not valid anymore.*/
+		DropAllShardSplitLeftOvers(sourceShardToCopyNode,
+								   shardSplitHashMapForPublication);
+
+		DropDummyShards();
```

Comment:
You can unify this with PG_FINALLY and not have to duplicat the call.

Reply:
We cannot unify this with FINALLY as the calls are different except for DropDummyShards().

DropDummyShards() will drop dummy shards if created.